### ISSUE TITLE
loss nan in Genomic Benchmark

### ISSUE THREAD

**Issue by multydoffer** [kuleshov-group/caduceus#34]
I am using the 470k released weight to finetune on genomic benchmark, but encounter nan loss on both ps and ph models. If the learing rate is set to 1e-3, the nan will occur in the 2nd epoch. What's going on? I've tried other size models, they can run smoothly
Furthermore, 1e-7 still produce nan. This is wired, cuz I have clip the gradient also. And I notice that in your paper you state that the lr is 1e-3 right?
And the lr is already set with cosine decay scheduler in the 1e-3 setting, which ends at 1e-4 after 10 epochs

**Comment by multydoffer** [comment]
Here is my customized code for running Genomic Benchmark, tokenizer is in other place, still aligned with char setting officially
`from transformers import AutoModelForMaskedLM, AutoTokenizer
import torch
import torch.nn as nn


class Caduceus(nn.Module):
    def __init__(self, mode="ph", size=1.9, out_dim=2, *args, **kwargs):
        super().__init__()
        # self.d_model = config.hidden_size
        self.mode = mode
        if mode=="ph":
            if size == 470:
                model_name = "kuleshov-group/caduceus-ph_seqlen-1k_d_model-118_n_layer-4_lr-8e-3"
            elif size==1.9:
                model_name = "kuleshov-group/caduceus-ph_seqlen-1k_d_model-256_n_layer-4_lr-8e-3"
            else:
                model_name = "kuleshov-group/caduceus-ph_seqlen-131k_d_model-256_n_layer-16"
            # tokenizer = AutoTokenizer.from_pretrained(model_name)
        else:
            if size == 470:
                model_name = "kuleshov-group/caduceus-ps_seqlen-1k_d_model-118_n_layer-4_lr-8e-3"
            elif size==1.9:
                model_name = "kuleshov-group/caduceus-ps_seqlen-1k_d_model-256_n_layer-4_lr-8e-3"
            else:
                model_name = "kuleshov-group/caduceus-ps_seqlen-131k_d_model-256_n_layer-16"
            # tokenizer = AutoTokenizer.from_pretrained(model_name) 
        self.caduceus = AutoModelForMaskedLM.from_pretrained(model_name, trust_remote_code=True)
        self.d_model = self.caduceus.config.d_model
        self.score = nn.Linear(self.d_model, out_dim, bias=False)
        self.pooling_strategy = "mean"
        # self.lm_head



    def forward(self, input_ids, position_ids=None, inference_params=None, state=None): # state for the repo interface
        if isinstance(input_ids, list):
            input_ids_tensor = input_ids[0]
            attention_mask = input_ids[1]
        else:
            input_ids_tensor = torch.tensor(input_ids)
            attention_mask = None
        if position_ids is not None:
            position_ids_tensor = position_ids
        else:
            position_ids_tensor = None

        outputs = self.caduceus.caduceus(
            input_ids=input_ids,
            inputs_embeds=None,
            output_hidden_states=None,
            return_dict=None,
        )
        sequence_output = outputs[0]
        if self.mode == "ps":
            sequence_output = torch.stack(
                [
                    sequence_output[..., :self.d_model],
                    torch.flip(sequence_output[..., self.d_model:], dims=[1, 2])
                 ],
                dim=-1
            )
        # Pool and get logits
        pooled_hidden_states = self.pool_hidden_states(sequence_output)
        # Potentially run `score` twice (with parameters shared) for conjoining
        if sequence_output.ndim == 4:  # bsz, seq_len, hidden_dim, 2 where last channel has the stacked fwd and rc reps
            logits_fwd = self.score(pooled_hidden_states[..., 0])
            logits_rc = self.score(pooled_hidden_states[..., 1])
            logits = (logits_fwd + logits_rc) / 2
        else:
            logits = self.score(pooled_hidden_states)
        return logits, None
        
    def pool_hidden_states(self, hidden_states, sequence_length_dim=1):
        """Pools hidden states along sequence length dimension."""
        if self.pooling_strategy == "mean":  # Mean pooling along sequence length dimension
            return hidden_states.mean(dim=sequence_length_dim)
        if self.pooling_strategy == "max":  # Max pooling along sequence length dimension
            return hidden_states.max(dim=sequence_length_dim).values
        if self.pooling_strategy == "last":  # Use embedding of last token in the sequence
            return hidden_states.moveaxis(hidden_states, sequence_length_dim, 0)[-1, ...]
        if self.pooling_strategy == "first":  # Use embedding of first token in the sequence
            return hidden_states.moveaxis(hidden_states, sequence_length_dim, 0)[0, ...]`

**Comment by yair-schiff** [comment]
What batch size are you using? I found that batch size also affected stability for these downstream tasks.

**Comment by multydoffer** [comment]
64

**Comment by multydoffer** [comment]
now I also tried 16 and 32, while the result still is nan. When the lr is 1e-10, it is not nan

**Comment by yair-schiff** [comment]
I think in our work we used batch size of 256 and LR of either 1e-3 or 2e-3 here and runs were generally stable.

Are you seeing the NaN instability for all tasks in the Genomics Benchmark or only a certain subset?

**Comment by multydoffer** [comment]
actually in all benchmark. The problem is the 470k weight, actually.

**Comment by multydoffer** [comment]
Are you fine running my customized caduceus with size set to 470?

**Comment by multydoffer** [comment]
Yeah, I just move some of the code in the sequenceclassification code out to my customized Caduceus class, other things are just the same.

**Comment by yair-schiff** [comment]
@multydoffer are you still having issues here?
